### PR TITLE
8282771 : Create test case for JDK-8262981

### PR DESCRIPTION
**Pull request by lawrence-andrew**
Following methods are covered in this testcase
getAccessibleAction()
getAccessibleActionCount()
doAccessibleAction(int direction) 

@shurymury

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8282771](https://bugs.openjdk.java.net/browse/JDK-8282771): Create test case for JDK-8262981

### Reviewers
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7734/head:pull/7734` \
`$ git checkout pull/7734`

Update a local copy of the PR: \
`$ git checkout pull/7734` \
`$ git pull https://git.openjdk.java.net/jdk pull/7734/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7734`

View PR using the GUI difftool: \
`$ git pr show -t 7734`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7734.diff">https://git.openjdk.java.net/jdk/pull/7734.diff</a>

</details>

### CODE REVIEW

**Comment by shurymury** [comment]
@azuev-java what do you think be a better location for this test? javax/swing/JSlider/8262981 or somewhere in javax/accessibility?

**Comment by azuev-java** [comment]
> @azuev-java what do you think be a better location for this test? javax/swing/JSlider/8262981 or somewhere in javax/accessibility?

I'd say let's group them up somewhere in accessibility folder so they can be ran easier as a bundle instead of selecting them from individual component test folders.

File: test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        robot.setAutoDelay(300);
+        robot.waitForIdle();
+
+        List<String> instLookAndFeels =
+                Arrays.stream(UIManager.getInstalledLookAndFeels())
+                        .map(UIManager.LookAndFeelInfo::getClassName).collect(toList());
+
+        for (String laf : instLookAndFeels) {
+            try {
+                invalidDecrementCountDownLatch = new CountDownLatch(1);
```
Comment (shurymury) [inline test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java:179]: these fields are accessed both from EDT and the main test thread

File: test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+                            "return AccessibleAction.DECREMENT");
+                }
+
+                // JSlider value is incremented
+                mouseAction(robot, incrementBtn);
+                if (!validIncrementCountDownLatch.await(30, TimeUnit.SECONDS)) {
+                    throw new RuntimeException("Failed to perform action on valid " +
+                            "Increment button");
+                }
+                if (jSliderInitialValue.get() != currentJSliderValue.get()) {
```
Comment (shurymury) [inline test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java:264]: This is  fundamentally unstable. 

You are clicking on the decrease button. The actual decrease which happened in the slider after the click may be more than you expect. Or less. 

The only reliable condition perhaps is that the value has decreased and even that may have a delayed effect, so it needs to be waited for. 

For example see  sanity/client/SwingSet/src/SliderDemoTest.java.

File: test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java (hunk 1 above)
Comment (lawrence-andrew) [reply test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java]: Fixed

File: test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java (hunk 2 above)
Comment (lawrence-andrew) [reply test/jdk/javax/swing/JSlider/8262981/JSliderAccessibleAction.java]: Fixed this by adding addChangeListener to JSlider and updating the currentJSliderValue.

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+                point[0].y + rect[0].height / 2);
+        robot.waitForIdle();
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+        robot.waitForIdle();
+    }
+
+    public static void main(String[] args) throws InterruptedException,
+            InvocationTargetException, AWTException {
+        JSliderAccessibleAction jSliderAccessibleAction =
```
Comment (shurymury) [inline test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java:282]: The class appears to have no state, why do you need an instance of it? Why not to make the methods static?

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                Arrays.stream(UIManager.getInstalledLookAndFeels())
+                        .map(UIManager.LookAndFeelInfo::getClassName).collect(toList());
+
+        for (String lookAndFeel : installedLookAndFeels) {
+            try {
+                invalidDecrementCountDownLatch = new CountDownLatch(1);
+                invalidIncrementCountDownLatch = new CountDownLatch(1);
+                validDecrementCountDownLatch = new CountDownLatch(1);
+                validIncrementCountDownLatch = new CountDownLatch(1);
+                currentJSliderValue = new AtomicInteger();
```
Comment (shurymury) [inline test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java:179]: if this is a volatile, than perhaps you do not need an atomic?

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    private static JButton incrementBtn;
+    private static JButton invalidDecrementBtn;
+    private static JButton invalidIncrementBtn;
+    private static final int INVALID_DECREMENT = 2;
+    private static final int INVALID_INCREMENT = -1;
+    private static final int VALID_DECREMENT = 1;
+    private static final int VALID_INCREMENT = 0;
+    private static volatile AtomicInteger currentJSliderValue;
+    private static volatile AtomicInteger jSliderInitialValue;
+    private static volatile AccessibleAction accessibleAction;
```
Comment (shurymury) [inline test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java:74]: This field seem to be only accessed from the EDT, similar to UI components above, so why is it volatile and the components are not?

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+                });
+            }
+        }
+    }
+
+    public static void mouseAction(Robot robot, JButton button) throws InterruptedException,
+            InvocationTargetException {
+        robot.waitForIdle();
+        final Point[] point = new Point[1];
+        final Rectangle[] rect = new Rectangle[1];
```
Comment (shurymury) [inline test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java:265]: finals here are redundant although do not hurt

**Review by shurymury — CHANGES_REQUESTED** [review]

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java (hunk 3 above)
Comment (lawrence-andrew) [reply test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java]: Fixed make the method to be static

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java (hunk 4 above)
Comment (lawrence-andrew) [reply test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java]: Fixed

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java (hunk 6 above)
Comment (lawrence-andrew) [reply test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java]: Fixed

File: test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java (hunk 5 above)
Comment (lawrence-andrew) [reply test/jdk/javax/accessibility/JSlider/AccessibleAction/JSliderAccessibleAction.java]: Fixed

**Comment by shurymury** [comment]
LGTM

**Pull request closed without merging**